### DISCUSSION TITLE
Add domain support to Blueprint

### DISCUSSION THREAD

**Discussion by fabswt** [pallets/flask#5413] [Q&A]
Currently, Blueprints support a `url_prefix` and a `subdomain` option. I would like to see support for a `domain` option.

I need to host the same codebase on two different sites, but with some of the routes only available to one domain and others only to the other, and roughly 3/4 of routes available on both sites.

I don't want to split the codebase into two because 95% of the code is the same. Checking for `request.host` on every route (even with a route decorator) is cumbersome. Having this at the level of Blueprints would make the most sense.

There's [a way to support multiple domains](https://gist.github.com/tachyondecay/a17960d5d8276387e7cf6982760e7ec2) but it does not allow for the sharing of routes.

**Comment by davidism** [comment]
I'm not clear what the request is here. You should already be able to do this with `host_matching=True`. For the 5% that app A needs, make a blueprint only for A's domain, same for the 5% that only B needs.

**Comment by davidism** [comment]
Maybe you're saying that `Blueprint()` does not take a `host=` parameter. But `bp.register()` _does_ take a `host` parameter. So you can still apply this to all routes on a blueprint without change, you just have to specify it at registration instead of at creation.

**Comment by fabswt** [comment]
I want some routes to be available to example.org only, others to somesite.com only, and some to both. Didn't find any example of how to do it with Blueprints.

What's `bp.register()`?

**Comment by davidism** [comment]
Sorry, meant `app.register_blueprint`. But you can do that. You can have any layout you want with blueprints. So make some that are shared, some that have `host=first_host`, and some that have `second_host`.

```python
a_bp = Blueprint()
b_bp = Blueprint()
shared_bp = Blueprint()

app.register_blueprint(a_bp, host="a")
app.register_blueprint(b_bp, host="b")
app.register_blueprint(shared_bp)
```

Something along those lines. Sorry, I don't really use this feature, so I don't have a ready made environment to spin up to answer your specific question.

**Comment by fabswt** [comment]
i have the following example:

```python
"""
attempt at an MVCE for Flask with domain blueprints (Flask 3.x)
python3 flaskapp-with-blueprint-example-1.py
"""

from flask import Blueprint, Flask, request


def create_app():
    app = Flask(__name__)

    first_bp = Blueprint('first', __name__)
    second_bp = Blueprint('second', __name__)
    shared_bp = Blueprint('shared', __name__)

    @first_bp.route('/')
    def first_home():
        return f"""route meant only for first.localhost"""

    @second_bp.route('/')
    def second_home():
        return f"""route meant only for second.localhost"""

    @shared_bp.route('/shared')
    def shared_page():
        return f"""this route should work with both domains"""

    app.register_blueprint(first_bp, host="first.localhost")
    app.register_blueprint(second_bp, host="second.localhost")
    app.register_blueprint(shared_bp)

    return app


if __name__ == "__main__":
    app = create_app()
    app.run(
        debug=True,
        host='0.0.0.0',
        port=6123,

        # if running it inside of my docker container:
        # ssl_context=(
        #     "/etc/apache2/ssl/SSLforMyHosts-certificate.pem",
        #     "/etc/apache2/ssl/SSLforMyHosts-key.pem",
        # ),
    )
    
"""adding these lines to your /etc/hosts file should be enough to get up and running:
127.0.0.1 first.localhost
127.0.0.1 second.localhost
"""
```



it runs, but not as expected:

- http://first.localhost:6123/ returns "route meant only for first.localhost" (ok)
- http://second.localhost:6123/ returns "route meant only for first.localhost" as well which is wrong. expected to get "route meant only for second.localhost".
- http://first.localhost:6123/shared returns "this route should work with both domains" (ok)
- http://second.localhost:6123/shared returns "this route should work with both domains" (ok)


i thought maybe i need this to set `app` instead:

```
app = Flask(__name__, host_matching=True, static_host='first.localhost')
```

but then every route becomes 404 Not Found.

i don't understand what `static_host` is. i guess it's the host used to serve static data but it's counter-intuitive that i should set it to a single value when working with multiple domains. likewise for `app.config['SERVER_NAME']` (i don't understand if i'm supposed to set it or not, either way setting it to 'first.localhost' did not help in the examples above.)


no offense but the documentation could be clearer. `host_matching` and domain are not mentioned once on the blueprints page (only subdomains are.)

**Reply by samuelhwilliams** [reply]
So you do need this line:

```
app = Flask(__name__, host_matching=True, static_host='first.localhost')
```

Importantly, you also need to add the port to the `host` declarations.

Eg:
```
    app.register_blueprint(first_bp, host="first.localhost:6123")
    app.register_blueprint(second_bp, host="second.localhost:6123")
```

---

I've modified the script myself (runs on port 8080) now to expand the examples:

```python
from flask import Blueprint, Flask, request


def create_app():
    app = Flask(__name__, host_matching=True, static_host='static.localhost:8080')

    first_bp = Blueprint('first', __name__)
    second_bp = Blueprint('second', __name__)
    shared_bp = Blueprint('shared', __name__)

    @first_bp.route('/')
    @first_bp.route('/first', host='first.localhost:8080')
    @first_bp.route('/overlap', host='first.localhost:8080')
    def first_home():
        return f"""route meant only for first.localhost"""

    @second_bp.route('/')
    @second_bp.route('/second', host='second.localhost:8080')
    @second_bp.route('/overlap', host='second.localhost:8080')
    def second_home():
        return f"""route meant only for second.localhost"""

    @shared_bp.route('/shared')
    def shared_page():
        return f"""this route should work with both domains"""

    app.register_blueprint(first_bp, host="first.localhost:8080")
    app.register_blueprint(second_bp, host="second.localhost:8080")
    app.register_blueprint(shared_bp)

    return app


if __name__ == "__main__":
    app = create_app()
    app.run(
        debug=True,
        host='0.0.0.0',
        port=8080,
    )
```

And here's what I see:

| Request URL | Response |
|---|---|
| http://first.localhost:8080/ | 404 |
| http://first.localhost:8080/first | `route meant only for first.localhost` |
| http://first.localhost:8080/overlap | `route meant only for first.localhost` |
| http://second.localhost:8080/ | 404 |
| http://second.localhost:8080/second | `route meant only for second.localhost` |
| http://second.localhost:8080/overlap | `route meant only for second.localhost` |
| http://localhost:8080/shared | 404 |
| http://127.0.0.1:8080/shared | 404 |
| http://first.localhost:8080/shared | 404 |
| http://second.localhost:8080/shared | 404 |

**Comment by tachyondecay** [comment]
@fabswt  I have been using host matching in production successfully for years. You can see how I’m doing it on [this public repo](https://github.com/tachyondecay/tachyondecay.net-mirror).

I did end up subclassing some of the blueprint-related classes to achieve two things. First, I can set a host on an entire blueprint instead of per-route (David's example above suggests this was already possible, so maybe I missed that when I was setting this up years ago); second, I have host matching on static endpoints (which it sounds like you want too).

You’re welcome to look at how I did that and borrow it, although I make no warranty to whether it’s “good.” So far it has worked for me.

If you have questions, please feel free to ask me here, or you can ping me on the Pallets Discord (same username).

**Comment by fabswt** [comment]
@davidism any chance of getting feedback as to whether:

a. my Minimal, Reproducible Example is wrong

or

b. Blueprints actually do not support domains out of the box.

?

---

@tachyondecay thank you for the comment, appreciated. I actually stumbled upon your repo while trying to get things to work. Still trying to get things working with out-of-the-box features and a minimal example, but will definitely revisit your code if that doesn't work out.

**Comment by fabswt** [comment]
for what it's worth, i just gave up. the minimal example i gave should be moved into an issue and/or the docs should be updated as they only mention subdomains and not domains.
